internal/datanode/compactor.go

```diff
@@ -0,0 +1,549 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datanode
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/storage"
+	"github.com/milvus-io/milvus/internal/types"
+
+	"github.com/milvus-io/milvus/internal/proto/commonpb"
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+	"github.com/milvus-io/milvus/internal/proto/etcdpb"
+	"github.com/milvus-io/milvus/internal/proto/schemapb"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	errCompactionTypeUndifined = errors.New("compaction type undefined")
+	errIllegalCompactionPlan   = errors.New("compaction plan illegal")
+	errTransferType            = errors.New("transfer intferface to type wrong")
+	errUnknownDataType         = errors.New("unknown shema DataType")
+)
+
+type iterator = storage.Iterator
+
+type compactor interface {
+	compact() error
+	getPlanID() UniqueID
+}
+
+type compactionTask struct {
+	downloader
+	uploader
+	compactor
+	Replica
+	flushManager
+	allocatorInterface
+
+	dc   types.DataCoord
+	plan *datapb.CompactionPlan
+}
+
+// check if compactionTask implements compactor
+var _ compactor = (*compactionTask)(nil)
+
+func newCompactionTask(
+	dl downloader,
+	ul uploader,
+	replica Replica,
+	fm flushManager,
+	alloc allocatorInterface,
+	dc types.DataCoord,
+	plan *datapb.CompactionPlan) *compactionTask {
+	return &compactionTask{
+		downloader:         dl,
+		uploader:           ul,
+		Replica:            replica,
+		flushManager:       fm,
+		allocatorInterface: alloc,
+		dc:                 dc,
+		plan:               plan,
+	}
+}
+
+func (t *compactionTask) getPlanID() UniqueID {
+	return t.plan.GetPlanID()
+}
+
+func (t *compactionTask) mergeDeltalogs(dBlobs map[UniqueID][]*Blob, timetravelTs Timestamp) (map[UniqueID]Timestamp, *DelDataBuf, error) {
```

Comment:
how do we handle the case when a pk is deleted twice?

Reply:
1. Deserializing of Deltalogs has no dup pk.
2. Not sure if delete will write 2 same pk delete into deltalogs

Reply:
This compaction will delete all the same pks in InsertLogs.

Reply:
> 1. Deserializing of Deltalogs has no dup pk.
> 2. Not sure if delete will write 2 same pk delete into deltalogs

Delete cannot delete a PK twice now.